include/cucumber-cpp/internal/utils/qtCapture.hpp

```diff
@@ -0,0 +1,182 @@
+/* Based on Sir Digby Chicken post from:
+ * http://stackoverflow.com/questions/5419356/redirect-stdout-stderr-to-a-string
+ * who granted his work on BSD license
+ * adjusted for Qt by konserw
+ */
+#if defined(WIN32) || defined(_WIN32) || defined(__WIN32) && !defined(__CYGWIN__)
+#define WINDOWS
+#include <io.h>
+#include <fcntl.h>
+#define popen _popen 
+#define pclose _pclose
+#define stat _stat 
+#define dup _dup
+#define dup2 _dup2
+#define fileno _fileno
+#define close _close
+#define pipe _pipe
+#define read _read
+#define eof _eof
+#else
+#include <errno.h>
+#endif
+
+#include <cstdio>
+#include <QtConcurrent/QtConcurrent>
+
+#ifndef STD_OUT_FD
+#define STD_OUT_FD (fileno(stdout))
+#endif
+#ifndef STD_ERR_FD
+#define STD_ERR_FD (fileno(stderr))
+#endif
+
+class qtCapture {
+public:
+    static void Init() {
+        // make stdout & stderr streams unbuffered
+        // so that we don't need to flush the streams
+        // before capture and after capture 
+        // (fflush can cause a deadlock if the stream is currently being 
+        QMutexLocker locker(&m_mutex);
+        setvbuf(stdout, NULL, _IONBF, 0);
+        setvbuf(stderr, NULL, _IONBF, 0);
+    }
+
+    static void BeginCapture() {
+        QMutexLocker locker(&m_mutex);
+        if (m_capturing)
+            return;
+
+        secure_pipe(m_pipe);
+        m_oldStdOut = secure_dup(STD_OUT_FD);
+        m_oldStdErr = secure_dup(STD_ERR_FD);
+        secure_dup2(m_pipe[WRITE], STD_OUT_FD);
+        secure_dup2(m_pipe[WRITE], STD_ERR_FD);
+        m_capturing = true;
+#ifndef WINDOWS
+        secure_close(m_pipe[WRITE]);
+#endif
+    }
+
+    static bool IsCapturing() {
+        QMutexLocker locker(&m_mutex);
+        return m_capturing;
+    }
+
+    static void EndCapture() {
```

Comment:
Why is this entire class a singleton? It seems to me like the restoring of file descriptors that you're doing here is something that you would want to do in a destructor.